lib/UndoTruncateToOddIntegerPass.cpp

```diff
@@ -56,44 +58,45 @@ struct UndoTruncateToOddIntegerPass : public ModulePass {
     }
 
     // This base case makes for easier recursion.
-    switch (bit_width) {
-    case 8:
-    case 16:
-    case 32:
-    case 64:
-      if (!isa<ZExtInst>(v))
-        return v;
-    default:
-      break;
-    }
+    if (bit_width == desired_bit_width && !isa<ZExtInst>(v))
+      return v;
 
+    auto desired_int_ty = IntegerType::get(v->getContext(), desired_bit_width);
     if (auto *ci = dyn_cast<ConstantInt>(v)) {
-      return ConstantInt::get(i32_, uint32_t(ci->getZExtValue()));
+      return ConstantInt::get(desired_int_ty, uint32_t(ci->getZExtValue()));
     }
     Value *result = nullptr;
     if (auto *trunc = dyn_cast<TruncInst>(v)) {
-      auto *operand = ZeroExtend(trunc->getOperand(0));
+      auto *operand = ZeroExtend(trunc->getOperand(0), desired_bit_width);
 
       // Now, and the extended version to keep the range of the output
       // restricted to the original bit width.
       result = BinaryOperator::Create(
           Instruction::And, operand,
           ConstantInt::get(
-              i32_, (uint32_t)APInt::getAllOnesValue(bit_width).getZExtValue()),
+              desired_int_ty,
+              (uint32_t)APInt::getAllOnesValue(bit_width).getZExtValue()),
           "", trunc);
     } else if (auto *zext = dyn_cast<ZExtInst>(v)) {
-      result = new ZExtInst(zext->getOperand(0), i32_, "", zext);
+      auto tmp = ZeroExtend(zext->getOperand(0), desired_bit_width);
+      auto operand_bit_width = tmp->getType()->getIntegerBitWidth();
+      if (isPowerOf2_32(operand_bit_width)) {
+        result = tmp;
+      } else {
+        result = new ZExtInst(zext->getOperand(0), desired_int_ty, "", zext);
```

Comment:
I don't quite understand what situation this code is handling. When is `result = tmp` not the right thing to do? Doesn't `ZeroExtend()` always produce a result with the desired bit width? I couldn't come up with a sequence of instructions that triggers the `else` clause.

Reply:
I've reworked this and added a few tests with mixed sizes that need extended (not all 32-bits).